External/FEXCore/include/FEXCore/Core/CoreState.h

```diff
@@ -8,8 +8,72 @@
 #include <cstddef>
 #include <stdint.h>
 #include <string_view>
+#include <type_traits>
 
 namespace FEXCore::Core {
+  // This is defined here instead of a helper because it is a massive footgun in behaviour.
+  // Don't use this instead of std::atomic unless you really know what you're doing.
+  // Increment and decrement can visibily tear and only uses relaxed atomic behaviour internally.
+  // In particular decrement and increment can tear if a signal is received half-way through
+  // and you need to be aware of that when using this.
+  //
+  // Primarily this ensure program ordering when signals are concerned.
+  //
+  // This casts its internal object to std::atomic internally so that compilers can't reorder operations around it.
+  // This is necessary with how it is used with deferring signals.
+  template<typename T>
+  class MoveableNonatomicRefCounter {
```

Comment:
Renamed this to `NonAtomicRefCounter` and used std::atomic directly.
This is because it was required to be moved.
`CreateDefaultCPUState` would create a CPUState object which would get moved in to position.
I have changed this function to `AssignDefaultThreadState` which instead will set the state directly in the newly created object rather than moving it around.

Reply:
What was the reason for changing it around like this? Avoiding two-phase initialization is a valid reason to provide moveability.
(IIRC my comment wasn't a critique against being moveable; it just seemed weird to highlight the notion of it being movable in the class name)

Reply:
Any update on this?

Reply:
It was renamed to `NonAtomicRefCounter`, and it just uses std::atomic now.
CPUState used to be initialized and then moved to the final location inside of `InternalThreadState` but I changed the code to initialize the CPUState directly inside of the `InternalThreadState` object instead.
So there shouldn't be anything else to do here.

Reply:
The two phase initialization was necessary because we couldn't move the thread logic after initialization happened since atomics can't be moved.
Removed the two stage initialization entirely. CPUState now just initializes its own state in its constructor.